### DISCUSSION TITLE
Use `Request` instead of Pydantic model in `POST`, but still document the pydanti model

### DISCUSSION THREAD

**Discussion by achempak-polymer** [fastapi/fastapi#11473] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Request
from pydantic import BaseModel

app = FastAPI()

class Foo(BaseModel):
  bar: str

@app.post("/test")
async def test(req: Request):
  body = await req.body().decode("utf-8")
  # do stuff with body
  parsed = Foo.model_validate_json(body)
  # do stuff with parsed
```

### Description

Is there a way to still document `Foo` as the request body for `/test` while still using `Request` instead of `Foo` as the input for `test()`? The reason we want to do this is that we're getting ClientDisconnect errors *while* the request is being validated and parsed into the pydantic model if we use it as a parameter for `test()`, before the logic inside `test()` even has a chance to execute. The only way to catch these errors is to manually parse the raw `Request` object. Without this, we get nebulous ASGI exception errors in our logs that aren't very useful, and make it hard to distinguish more critical ASGI errors that might be due to memory leaks, for example.

### Operating System

Linux, macOS

### FastAPI Version

0.109.1

### Pydantic Version

2.3.0

### Python Version

3.11

**Comment by n8sty** [comment]
I'm not aware of a way to achieve exactly what you're asking but you may want to try to switch `Foo` to be a `dataclass` instead of a Pydantic model since the former does significantly less validation.

**Comment by YuriiMotov** [comment, marked as answer]
You can declare every path operation twice, but disable including in openapi schema for one of them. Only first operation will be executed
```
from fastapi import FastAPI, Request
from pydantic import BaseModel

app = FastAPI()


class Item(BaseModel):
    name: str


# This operation will NOT be included in openapi schema, but it will be executed
@app.post("/", include_in_schema=False)
async def route_real(request: Request):
    print("real")
    body = (await request.body()).decode("utf-8")
    parsed = Item.model_validate_json(body)
    print(parsed.model_dump())

# This operation will be included in openapi schema, but it will NOT be executed
@app.post("/")
def route_fake(item: Item):
    raise NotImplementedError()
```

If you perform a post request to "/" it will print "real" in console, but in the documentation you will see documentation for `route_fake` endpoint. You can rename function names as you want. You can even use the same name for both of them.

But I'm not sure if this is a good approach. I assume that this behavior may be changed in the future..

**Comment by sehraramiz** [comment]
one way to solve your problem is to define openapi specs separately with [custom openapi schema on routers](https://fastapi.tiangolo.com/advanced/path-operation-advanced-configuration/#custom-openapi-path-operation-schema). 

```python
from fastapi import FastAPI, Request
from pydantic import BaseModel

app = FastAPI()


class Foo(BaseModel):
    bar: str


@app.post(
    "/test",
    openapi_extra={
        "requestBody": {
            "content": {
                "application/json": {
                    "schema": Foo.schema()
                }
            },
            "required": True,
        },
    },
)
async def test(req: Request):
    body = await req.body()
    body = body.decode("utf-8")
    parsed = Foo.model_validate_json(body)
```

**Reply by nextmat** [reply]
An update on this approach. In Pydantic v2 `.schema()` has been renamed to [`.model_json_schema()`](https://docs.pydantic.dev/latest/concepts/json_schema/).

You will now want to use something like:

```py
@app.post(
    "/test",
    openapi_extra={
        "requestBody": {
            "content": {
                "application/json": {
                    "schema": Foo.model_json_schema(
                        ref_template="#/components/schemas/{model}"
                    )
                }
            },
            "required": True,
        },
    },
  },
)
```

For me the [`ref_template`](https://docs.pydantic.dev/latest/concepts/json_schema/#customizing-the-refs-in-json-schema) argument is required to get Pydantic to put the generated schema in the right place for OpenAPI.

**Comment by achempak-polymer** [comment]
Thanks both of you guys! These both work - I ended up going with @YuriiMotov 's approach since that one guarantees the schema will be the exact same as what I had before.
